### ISSUE TITLE
Nonce increment issue

### ISSUE THREAD

**Issue by SpasiboKojima** [MetaMask/metamask-mobile#4977]
**Describe the bug**
Submitting transaction from the same address that was just used for tx on desktop gives error, i.e. nonce doesn't get calculated on transaction prepare. Even reloading page doesn't help, only killing the app or manually incrementing nonce fixes the error.
It increments correctly only the first time you submit a tx from mobile after desktop, the second time it doesn't.
The same thing seems to reported in #4744  but wasn't properly described, I think.
What's also interesting is that the same thing happens on desktop Metamask but only visually, it doesn't give any errors and when you press confirm on tx it goes with correctly incremented nonce.

**Screenshots**

https://user-images.githubusercontent.com/34808650/189852818-c29c8313-d20c-401a-bda7-eaad30c19042.mov

![Screenshot_20220912-092901_MetaMask](https://user-images.githubusercontent.com/34808650/189854060-7fc8e377-d284-4fcd-ad7e-03788b69a76c.jpg)

**To Reproduce**
_Steps to reproduce the behavior_
1. Open a page with a button that creates some transaction on both mobile and desktop;
2. Push the button and confirm the transaction on desktop, wait for confirmation popup from MM;
3. Push the button and confirm the transaction on mobile, notice correct incremented nonce, wait for confirmation popup from MMM;
4. Repeat point 2;
5. Trying to repeat point 3 will give incorrect nonce and an error

**Expected behavior**
Nonce is correctly incremented ideally on confirmation screen or after confirming like on desktop MM.

**Smartphone (please complete the following information):**
 - Device: any
 - OS: any iOS and Android
 - App Version 5.6.0

-------------------------------------------------------------
_to be added after bug submission by internal support / PM_
**Severity**

- How critical is the impact of this bug on a user?
- Add stats if available on % of customers impacted
- Is this visible to all users?
- Is this tech debt?

**Comment by bschorchit** [comment]
I was able to reproduce this, but only after turning on "Customize transaction nonce" in settings. If the toggle is off the transaction will go through as normal.

**Comment by SpasiboKojima** [comment]
> I was able to reproduce this, but only after turning on "Customize transaction nonce" in settings. If the toggle is off the transaction will go through as normal.

Weird, in my case we were having errors before turning this on, turning it on only made clear the origin of the error.

https://user-images.githubusercontent.com/34808650/190572842-bedb40f3-2144-467a-aa9e-440a2097736d.mov

**Comment by tfalencar** [comment]
I'm also having this issue. When pairing Metamask mobile with a local hardhat instance, the following is shown: 
```
eth_sendRawTransaction
  Nonce too low. Expected nonce to be 14 but got 13.
```
It seems either enabling or disabling "custom nonce" setting doesn't make any difference on mobile (There is no input field to add a custom nonce number even when the setting is enabled).

This problem makes it hard to test things end-to-end with metamask..
Reseting the account doesn't help either (although on the browser extension it does clear the nonce cache and provides a workaround  - albeit annoying).

**Comment by tfalencar** [comment]
> I was able to reproduce this, but only after turning on "Customize transaction nonce" in settings. If the toggle is off the transaction will go through as normal.

The happens when there is a mismatch between what the wallet thinks the state of the network is, versus the state of the network itself. 
This is more common when using the wallet during development of dapps, where the blockchain state is "rewind" by restarting / spawning a local environment (nonce value is reset to a lower number).
Probably metamask queries the network once for this information, then persist it instead of looking up every time ?

**Comment by seaona** [comment]
@tfalencar thanks for your summary! I've double checked and also encountering the same issues.
- For  "Reset the account" function not working, I've opened a separate issue here https://github.com/MetaMask/metamask-mobile/issues/5788
- For issue nº2 , we have an open ticket for the Approve screen, which is the only screen where you cannot Edit the nonce at t he moment https://github.com/MetaMask/metamask-mobile/issues/5395
- For issue nº3 we have a PR for fixing this here https://github.com/MetaMask/metamask-mobile/pull/5718

cc @bschorchit

**Comment by bschorchit** [comment]
Let's use this ticket to focus on issue number 3 for which @blackdevelopa might have already started a PR for here https://github.com/MetaMask/metamask-mobile/pull/5718 as @seaona pointed out.

For issues 1 and 2, we'll track them in their respective issues.

**Comment by bschorchit** [comment]
Closing this as https://github.com/MetaMask/metamask-mobile/pull/5718 has been merged. A fix will be available on app version 6.5
